File: C++/2-Medium/00255._验证二叉搜索树的前序遍历序列/00255._验证二叉搜索树的前序遍历序列.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

// 验证是否为二叉搜索树的前序遍历序列
bool verifyPreorder(vector<int>& preorder) {
    stack<int> stk;
    int lower_bound = INT_MIN; // 设置当前节点值可以取的最小值
    for (int num : preorder) {
        // 如果当前元素小于lower_bound，则不是有效的前序遍历序列
        if (num < lower_bound) return false;
        // 当前元素大于栈顶元素，更新lower_bound
        while (!stk.empty() && num > stk.top()) {
            lower_bound = stk.top(); // 更新为右子树的新的最小值
            stk.pop();
        }
        stk.push(num); // 将当前元素压入栈中
    }
    return true; // 遍历完毕，未发现违反规则的情况
}

int main() {
    // 示例
    vector<int> preorder1 = {5, 2, 6, 1, 3};
    vector<int> preorder2 = {5, 2, 1, 3, 6};
    cout << "Example 1: " << (verifyPreorder(preorder1) ? "true" : "false") << endl;
    cout << "Example 2: " << (verifyPreorder(preorder2) ? "true" : "false") << endl;
    return 0;
}

File: Python/2-Medium/00255._验证二叉搜索树的前序遍历序列/00255._验证二叉搜索树的前序遍历序列.py
from typing import List

def verify_preorder(preorder: List[int]) -> bool:
    """
    验证给定的序列是否可以表示某个二叉搜索树的前序遍历序列。
    
    Args:
    - preorder: List[int] - 一个整数数组，表示树的前序遍历序列。
    
    Returns:
    - bool - 如果序列是某个二叉搜索树的前序遍历，则返回True；否则返回False。
    
    Example:
    >>> verify_preorder([5,2,1,3,6])
    True
    >>> verify_preorder([5,2,6,1,3])
    False
    """
    # 初始化一个栈用于模拟前序遍历的过程
    stack = []
    # 初始化一个变量用于记录上一个根节点的值
    lower_bound = float('-inf')
    
    for value in preorder:
        # 如果当前值小于上一个根节点的值，返回False
        if value < lower_bound:
            return False
        # 当栈不为空，且栈顶元素小于当前值时，弹出栈顶元素
        while stack and stack[-1] < value:
            lower_bound = stack.pop()
        # 将当前值压入栈中
        stack.append(value)
    
    # 如果遍历完整个序列都没有发现问题，返回True
    return True

# 测试代码
test_cases = [
    ([5,2,1,3,6], True),
    ([5,2,6,1,3], False)
]

# 运行测试案例
for i, (preorder, expected) in enumerate(test_cases, 1):
    result = verify_preorder(preorder)
    print(f"测试案例 {i}: 预期结果={expected}, 实际结果={result}, {'通过' if result == expected else '失败'}")